contributors/design-proposals/selector-generation-for-all.md

```diff
@@ -0,0 +1,71 @@
+Selector Generation for Deployments, ReplicaSet, DaemonSet, StatefulSet and ReplicationController
+=============
+
+# Goals
+Make selector easy to use and less error-prone for Deployments, ReplicaSet, DaemonSet, StatefulSet and ReplicationController. Make `kubectl apply` work well with selector.
+
+# Problem Description
+The field `spec.selector` of Deployments, ReplicaSet, DaemonSet are defaulted from `spec.template.metadata.labels`, if it is unspecified.
+And there is validation to make sure `spec.selector` always selects the pod template.
+The defaulting of selector may prevent from `kubectl apply` from working when updating the `spec.selector` field. e.g. [kubernetes/kubernetes#26202 (comment)](https://github.com/kubernetes/kubernetes/issues/26202#issuecomment-221421254)
+
+# Proposed changes
+
+The [Selector Generation](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/selector-generation.md) works well in Job to make sure non-overlapping. We can use this pattern for Deployments, ReplicaSet, DaemonSet, StatefulSet and ReplicationController.
+If `spec.selector` is not derived from `spec.template.metadata.labels`, `kubectl apply` will work well for selectors.
+
+## API
+
+The change will be similar to [the changes to Job](https://github.com/kubernetes/kubernetes/blob/master/docs/design/selector-generation.md#api). But the defaulting is different from Job's.
+
+`extension/v1beta1 Deployment|ReplicaSet|DaemonSet`, `apps/v1beta1 StatefulSet` and `v1 ReplicationController` change as follows.
+
+Add field `spec.manualSelector`. It controls if using automatic generated selectors.
```

Comment:
I am not sure I like the additional field. I suggest we always default a key - is there a backwards-compatibility concern here?